jax/interpreters/sharded_jit.py

```diff
@@ -279,19 +188,84 @@ def _sharded_call_impl(fun: lu.WrappedFun, *args, **params):
 sharded_call = partial(core.call_bind, sharded_call_p)
 sharded_call_p.def_custom_bind(sharded_call)
 sharded_call_p.def_impl(_sharded_call_impl)
-xla.call_translations[sharded_call_p] = _sharded_jit_translation_rule
 
 
-def sharded_jit(fun, partitions):
-  if xb.get_backend().platform != "tpu":
-    logging.warning("sharded_jit only works on TPU")
+class PartitionSpec(tuple):
+  """Tuple of integer specifying how a value should be partitioned.
+
+  Each integer corresponds to how many ways a dimension is partitioned. We
+  create a separate class for this so JAX's pytree utilities can distinguish it
+  from a tuple that should be treated as a pytree.
+  """
+  def __new__(cls, *partitions):
+    return tuple.__new__(PartitionSpec, partitions)
+
+  def __repr__(self):
+    return "PartitionSpec%s" % tuple.__repr__(self)
+
+
+def sharded_jit(fun: Callable, in_parts, out_parts):
+  """Like ``jit``, but partitions ``fun`` across multiple devices.
+
+  WARNING: this feature is still under active development! It may not work well,
+  and may change without warning!
+
+  `sharded_jit` sets up ``fun`` for just-in-time compilation with XLA, but
+  unlike ``jit``, the compiled function will run across multiple devices
+  (e.g. multiple GPUs or multiple TPU cores). This is achieved by spatially
+  partitioning the data that flows through the computation, so each operation is
+  run across all devices and each device runs only a shard of the full
+  data. (Some data can optionally be replicated, which is sometimes more
+  efficient for small arrays when combined with larger spatially-partitioned
+  arrays.) Communication between devices is automatically inserted as necessary.
+
+  ``sharded_jit`` can be useful if the jitted version of ``fun`` would not fit
+  in a single device's memory, or to speed up ``fun`` by running each operation
+  in parallel across multiple devices.
+
+  Note: ``sharded_jit`` is currently available on TPU only!
+
+  Args:
+    fun: Function to be jitted.
+    in_parts: The input partitions, i.e. how each argument to ``fun`` should be
+      partitioned or replicated. This should be a PartitionSpec indicating into
+      how many partitions each dimension should be sharded, None indicating
+      replication, or (nested) standard Python containers thereof. For example,
+      ``in_parts=PartitionSpec(2,1)`` means all arguments should be partitioned
+      over two devices across the first dimension;
+      ``in_parts=(PartitionSpec(2,2), PartitionSpec(4,1), None)`` means the
+      first argument should be partitioned over four devices by splitting the
+      first two dimensions in half, the second argument should be partitioned
```

Comment:
I think I'd support a third kind of `in_part`: an integer, meaning partition that numbered axis `n_partitions` ways (especially if that's the 90% use case).

Reply:
(Although in that case I'd maybe want the kwargs to be `in_axes` and `out_axes` 🙂 )

Reply:
Note there is no `n_partitions` argument yet :P I think this is a good idea though. One could also write a wrapper function (maybe? It gets confusing since you currently need to know the rank of the arguments, although we could relax that too).